sdk/iot/core/src/az_iot_hub_client.c

```diff
@@ -1,12 +1,111 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+#include <stdint.h>
+
 #include "az_iot_hub_client.h"
+#include <az_precondition_internal.h>
+#include <az_result.h>
 #include <az_span.h>
 
 #include <_az_cfg.h>
 
+static const uint8_t hub_client_forward_slash = '/';
+static const uint8_t hub_client_param_separator = '&';
+
+static const az_span hub_client_api_version = AZ_SPAN_LITERAL_FROM_STR("?api-version=2018-06-30");
+
 AZ_NODISCARD az_iot_hub_client_options az_iot_hub_client_options_default()
 {
   return (az_iot_hub_client_options){ .module_id = AZ_SPAN_NULL, .user_agent = AZ_SPAN_NULL };
 }
+
+AZ_NODISCARD az_result az_iot_hub_client_init(
+    az_iot_hub_client* client,
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_iot_hub_client_options const* options)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(iot_hub_hostname, 1, false);
+  AZ_PRECONDITION_VALID_SPAN(device_id, 1, false);
+  AZ_PRECONDITION_NOT_NULL(options);
+
+  client->_internal.iot_hub_hostname = iot_hub_hostname;
+  client->_internal.device_id = device_id;
+  client->_internal.options.module_id = options->module_id;
+  client->_internal.options.user_agent = options->user_agent;
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_iot_hub_client_user_name_get(
+    az_iot_hub_client const* client,
+    az_span mqtt_user_name,
+    az_span* out_mqtt_user_name)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_user_name, 0, false);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_user_name);
+
+  const az_span* const module_id = &(client->_internal.options.module_id);
+  const az_span* const user_agent = &(client->_internal.options.user_agent);
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, client->_internal.iot_hub_hostname, &mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(mqtt_user_name, hub_client_forward_slash, &mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, client->_internal.device_id, &mqtt_user_name));
+  AZ_RETURN_IF_FAILED(
+      az_span_append_uint8(mqtt_user_name, hub_client_forward_slash, &mqtt_user_name));
+
+  if (az_span_length(*module_id) > 0)
+  {
+    AZ_RETURN_IF_FAILED(az_span_append(mqtt_user_name, *module_id, &mqtt_user_name));
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(mqtt_user_name, hub_client_forward_slash, &mqtt_user_name));
+  }
+
+  AZ_RETURN_IF_FAILED(
+      az_span_append(mqtt_user_name, hub_client_api_version, &mqtt_user_name));
+
+  if (az_span_length(*user_agent) > 0)
+  {
+    AZ_RETURN_IF_FAILED(
+        az_span_append_uint8(mqtt_user_name, hub_client_param_separator, &mqtt_user_name));
+    AZ_RETURN_IF_FAILED(az_span_append(mqtt_user_name, *user_agent, &mqtt_user_name));
+  }
+
+  *out_mqtt_user_name = mqtt_user_name;
+
+  return AZ_OK;
+}
+
+//[Format without module id] {device_id}
+//[Format with module id] {device_id}/{module_id}
```

Comment:
sorry, but lets remove this as header file should be good enough.

Reply:
woops yea removed the other one. will remove